Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs

```diff
@@ -0,0 +1,200 @@
+            Assert.NotNull(registration);
+            Assert.AreEqual(ServiceLifetime.Singleton, registration.Lifetime);
+        }
+
+        [When(@"I request an instance of the OpenApi host")]
+        public void WhenIRequestAnInstanceOfTheOpenApiHost()
+        {
+            ServiceProvider provider = this.scenarioContext.Get<ServiceProvider>();
+            IOpenApiHost<HttpRequest, IActionResult> host = provider.GetRequiredService<IOpenApiHost<HttpRequest, IActionResult>>();
+
+            this.scenarioContext.Set(host);
+        }
+
+        [Then(@"the HalDocumentMapper for resource type has configured its links")]
+        public void ThenTheHalDocumentMapperForResourceTypeHasConfiguredItsLinks()
+        {
+            ServiceProvider provider = this.scenarioContext.Get<ServiceProvider>();
+            PetHalDocumentMapper mapper = provider.GetRequiredService<PetHalDocumentMapper>();
+            Assert.IsTrue(mapper.LinkMapConfigured);
+        }
+
+        [Then(@"the HalDocumentMapper for resource and context types has configured its links")]
+        public void ThenTheHalDocumentMapperForResourceAndContextTypesHasConfiguredItsLinks()
+        {
+            ServiceProvider provider = this.scenarioContext.Get<ServiceProvider>();
+            PetHalDocumentMapperWithContext mapper = provider.GetRequiredService<PetHalDocumentMapperWithContext>();
+            Assert.IsTrue(mapper.LinkMapConfigured);
+        }
+
+        [Then(@"the exception of type ""(.*)"" is mapped to response code ""(.*)""")]
+        public void ThenTheExceptionOfTypeIsMappedToResponseCode(string exceptionType, int statusCode)
+        {
+            IOpenApiExceptionMapper exceptionMapper = this.scenarioContext.Get<ServiceProvider>().GetRequiredService<IOpenApiExceptionMapper>();
+
+            // The only way to tell if it's been mapped without reflecting into the guts of the mapper is to try and register a new mapper for
+            // the given type/status code... even with this we need a little bit of reflection to invoke the method.
+            var type = Type.GetType(exceptionType);
+            MethodInfo mapMethod = exceptionMapper.GetType().GetMethods().First(x => x.Name == "Map" && x.GetGenericArguments().Length == 1);
+            MethodInfo createdMapMethod = mapMethod.MakeGenericMethod(type);
+
+            try
+            {
+                createdMapMethod.Invoke(exceptionMapper, new object[] { statusCode, null });
+
+                Assert.Fail($"Exception of type '{exceptionType}' was not registered");
+            }
+            catch (TargetInvocationException ex)
+            {
+                // The exception will have been wrapped in a TargetInvocationException because we invoked via reflection, so we need to
+                // see if the inner exception was what we expected.
+                if (!(ex.InnerException is ArgumentException))
+                {
+                    throw;
+                }
+            }
+        }
```

Comment:
This test (the one above this comment) has baffled me. Its name implies to me that, broadly speaking, we feed an exception into the system, and then we want to check that a particular status code will emerge from somewhere else in the system. I don't actually see where that check is happening—you don't seem to be asking the question "Is this the status code we expect?" Instead, on the face of it you seem to be passing the status code into the code under test.

My gut feeling here is that this is another scenario where we should really be testing that the outcome is what we want, but it seems like what we're actually doing is checking things slightly further upstream, and just presuming that if we set everything in motion in the right way, we'll get the outcome we want. This might be more fragile, and it's certainly harder to understand: I don't yet understand how this test tests what it says it tests.

What I _think_ I want (although it's possible I'd change my mind once I saw what was involved in practice) is to have something that actually throws the exception, invoke that in exactly the way Menes normally would and to check the outcome. After all, the point of these tests is to ask "Is the out of the box behaviour correct when we've configured this in the default way?" We're not shining a light on detailed little corner cases in the exception handling here, we're just verifying that the basic cases operate as intended, right? And in particular, the broader question being asked by this particular set of tests is "Does everything work right after we've configured it?" Those feel like exactly the cases where you do actually want to involve quite a lot of the pieces in a test, because one of the main things configuration influences is how all the moving parts get plugged together. The overall effect is the thing we're interested in at this point, I think.

Reply:
So, I think this is a combination of bad naming and (potentially) missing documentation.

I'm not at all convinced that we should be testing the overall way Menes handles exceptions. These tests are supposed to be about initialisation rather than what happens when operations are invoked and throw exceptions. Would you make the same argument for the auditing specs in this feature file? I.e. that in order to know auditing is configured in the way it should be, we should be calling into the Menes pipeline and ensuring that the result is logged?

I'm not saying I think we shouldn't have those tests somewhere - but I think there's still value in having these tests as they are (albeit with better naming). There'd ideally be a whole separate feature file for Menes exception handling (and one for auditing) that effectively documented how these things work. But the fact that specific exception handling behaviour is configured as part of this particular way of setting Menes up is separate from that.

And we definitely need to decide the terminology we want to use for this exception handling functionality. At the moment, to be consistent with the naming in the code, I'd rename this scenario to `OpenApi host initialisation maps standard Menes exception types to their corresponding HTTP status codes`.

Reply:
> These tests are supposed to be about initialisation rather than what happens when operations are invoked and throw exceptions

Speaking as the named entity at the top of the feature file ("a developer" who wants "to be able to add Menes services and related components to my service collection"), I don't actually care a jot about how these configuration details effect the default behaviour. What I care about is the default behaviour. In fact I'd go as far as to say that one of the main points of having the various default configuration settings is to enable me to remain happily oblivious to how this aspects of configuration even work. I only want to have to think in terms of the relevant config settings if I want to change something.

So in my role as the alleged beneficiary of the feature file, I'd say that they are about how certain aspects of Menes behave out of the box. The key questions, then, are of the form: if I throw an `OpenApiBadRequestException` will I get a 400 response ? That's a condition and outcome that I care about. Config settings are basically irrelevant.

I've been thinking about this some more, and I've realised that *the heart of the matter* (for me) is this: suppose we changed the way the specified default behaviour was achieved, so that you get it through a total absence of config settings, and instead Menes just did the relevant thing by default unless you add config to change the behaviour. From the perspective of "default behaviour" that's an implementation detail. If I have a test that says "This is how this bit works by default" I wouldn't expect to need to change that test to handle this change from "defaults behaviour arises from automatically-supplied default config" to "default behaviour arises from how the system behaves when there's no config".

> Would you make the same argument for the auditing specs in this feature file? 

The only reason I didn't is that the tests stipulate that the auditing goes "to console" at which point this would then necessarily become a whole-process test. So I'm prepared to let that one go purely on pragmatic grounds, but I'll go on record as saying that it's a necessary evil rather than a lovely thing.

Having said all of that, we have made the decision that defaults are effected by supplying some initial configuration, and so it is arguably valid to have a test that just checks that this initial configuration looks like it's supposed to. I prefer the idea of a test that actually checks the behaviour because it seems inherently more robust, and a better expression of what I'm actually looking to get out of this as a developer using Menes, but if it's significantly more complex and the benefits are marginal, then deciding not to do that might be OK, in which case all we need to do is make it more clear what the test is actually checking for.

Reply:
Possibly the disconnect here then is that I am viewing these more as unit tests, ensuring that a specific bit of the code (initialisation) behaves as expected. I think your concern would be met by adding integration tests that describe the expected out of the box behaviour, as well as how it can be modified during initialisation.

Which is effectively me saying "I'd like to leave these tests as they are" and we should add a work item to cover creating specs that describe and test the out of the box behaviour. Are you happy with that approach?